File: 0131-palindrome-partitioning/0131-palindrome-partitioning.swift
class Solution {
    func partition(_ s: String) -> [[String]] {
        var result: [[String]] = []
        var currentList: [String] = []
        
        backtrack(s, &result, &currentList, 0)
        
        return result
    }
    
    func backtrack(_ s: String, _ result: inout [[String]], _ currentList: inout [String], _ start: Int) {
        if start == s.count {
            result.append(currentList)
            return
        }
        
        for end in start..<s.count {
            if isPalindrome(s, start, end) {
                let substring = String(s[s.index(s.startIndex, offsetBy: start)...s.index(s.startIndex, offsetBy: end)])
                currentList.append(substring)
                backtrack(s, &result, &currentList, end + 1)
                currentList.removeLast()
            }
        }
    }
    
    func isPalindrome(_ s: String, _ left: Int, _ right: Int) -> Bool {
        var leftIndex = left
        var rightIndex = right
        
        while leftIndex < rightIndex {
            if s[s.index(s.startIndex, offsetBy: leftIndex)] != s[s.index(s.startIndex, offsetBy: rightIndex)] {
                return false
            }
            leftIndex += 1
            rightIndex -= 1
        }
        
        return true
    }
}

File: 0131-palindrome-partitioning/0131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> [[str]]:
        def backtrack(i: int = 0):
            nonlocal ret, temp
            if i == len(s):
                ret.append(list(temp))
                return
            for length in range(len(s) - i):
                built_string = s[i:i + length + 1]
                if self.is_palindrome(built_string):
                    temp.append(built_string)
                    backtrack(i + length + 1)
                    temp.pop()

        temp = []
        ret = []
        backtrack()
        return ret

    def is_palindrome(self, string: str) -> bool:
        return string == string[::-1]
